Sharpen up descendant ordering terminology (#221)
ietf-wg-jsonpath/draft-ietf-jsonpath-base: draft-ietf-jsonpath-base.md

                           )
~~~~

> Note that there is no "bald" descendant selector (`..` on its own).

Reviewer: I'm not sure we need to introduce that term (even though, informally, I find it amusing).

```suggestion
Note that `..` on its own is not a valid selector.
```

Reviewer: Not a good suggestion?

Author: Whoops! Yes, it was a good suggestion, thanks. Since it's small, I'll piggy-back it on another PR.

Author: See https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/pull/214/commits/c7c245a95302c5e792d34e0ba4c6c9de10e431b2

#### Semantics
{: unnumbered}

> A `descendant-selector` selects certain descendants of a node:

Reviewer 2: Does `A` Indicate `any` or `for every`?

Author: I changed "The" to "A" because there are multiple descendant selectors. It carries the connotation of "any".

* the `..<name>` form (and the `..[<index>]` form where `<index>` is a `quoted-member-name`) selects those descendants that are member values of an object with the given member name.
* the `..[<index>]` form, where `<index>` is an `element-index`, selects those descendants that are array elements with the given index.
* the `..[<slice-index>]` form selects those descendants that are array elements selected by the given slice.
* the `..[<filter>]` form selects those descendants that are array elements or object values selected by the given filter.
* the `..[*]` and `..*` forms select all the descendants.

> An _array-sequenced preorder_ of the descendants of a node is a sequence of all the descendants in which:

Reviewer 2: Does  `An` Indicate `any` or `for every`?

#### Semantics
{: unnumbered}

A `descendant-selector` selects certain descendants of a node:

* the `..<name>` form (and the `..[<index>]` form where `<index>` is a `quoted-member-name`) selects those descendants that are member values of an object with the given member name.
* the `..[<index>]` form, where `<index>` is an `element-index`, selects those descendants that are array elements with the given index.
* the `..[<slice-index>]` form selects those descendants that are array elements selected by the given slice.
* the `..[<filter>]` form selects those descendants that are array elements or object values selected by the given filter.
* the `..[*]` and `..*` forms select all the descendants.

> An _array preorder_ of the descendants of a node is a sequence of all the descendants in which:

Reviewer: Hmm, the "preorder" part is universal and not just for arrays.  So qualifying "preorder" with "array" doesn't strike me as helpful.  Simply calling it "preorder" is fine, but I wanted to add an adjective that indicates that this conventional order is one of many possible (JSON objects), which the implementation is free to choose, and that it is the order we have chosen for JSONPath.

Author: Yeah, I was trying to capture an additional constraint on the order: it is both a preorder (parents before children) and arrays are traversed in array order (not implied by preorder).

So I tried to introduce a new term "array preorder" which is a special kind of preorder.

I'm not sure about an adjective conveying the lack of ordering of objects. Even if we found a suitable adjective, we'd still need to build in the array condition, ideally into the term, somehow.

Reviewer: After some more thinking, I don't get a better result than "array-ordered pre-order".  Maybe  bit redundant and wordy, but we are not using this term all over the place.

Author: That's certainly explicit, but the ordered/order part doesn't sound good. Do any of the following grab you?

* array-based preorder
* arraywise preorder
* array-respecting preorder
* arrayful preorder
* linear preorder
* vector preorder

Reviewer: Surprisingly hard.

* array-sequenced pre-order
* array-guided pre-order

Author: "array-sequenced preorder" seems to be the best term so far. (I think preorder is usually unhyphenated. See, for example, [wikipedia](https://en.wikipedia.org/wiki/Preorder).)

> * nodes of an array appear in array order, and

Reviewer: ```suggestion
* nodes of an array, immediately followed by their descendants appear in array order, and
```

Need to say the same for objects and their descendants.

> * nodes appear before their children.

Reviewer: hmm, this doesn't actually say depth first.

..* on
[1, 2, 3, [4, 5], 6]

could be shown as [1, 2, 3, [4, 5], 6, 4, 5] and satisfy the two rules.

Author: Good point. I'll address that shortly.
